buffer.go

```diff
@@ -0,0 +1,243 @@
+/*
+The Cache have RenameFrom : oldPath and RenameTo: newPath events cached.
+The time that an event was added is kept in timers cache. It's circular array.
+Every 100 miliseconds expired items are evicted from cache and if there is no match for them CREATE or DELETE event is created.
+
+
+*/
+
+package fsevents
+
+import (
+	"sync"
+	"time"
+)
+
+const (
+	MaxCapacity = 2000
+)
+
+type TimerCache struct {
+	Values               []TimerCacheItem
+	NumberOfItems        int
+	Capacity             int
+	TimerHead, TimerTail int
+}
+type Cache struct {
+	RenameFrom map[uint64]Event
+	RenameTo   map[uint64]Event
+	timers     TimerCache
+
+	cacheLock sync.RWMutex
+
+	ProcessedEvents chan []Event
+	done            chan bool
+}
+type TimerCacheItem struct {
+	T  time.Time
+	ID uint64
+}
+
+func (tc *TimerCache) moveHead() {
+	tc.TimerHead = (tc.TimerHead + 1) % tc.Capacity
+	tc.NumberOfItems--
+}
+func (tc *TimerCache) Add(t time.Time, id uint64) {
+	tc.Values[tc.TimerTail] = TimerCacheItem{t, id}
+	tc.NumberOfItems++
+}
+func (tc *TimerCache) Next() {
+	tc.TimerTail = (tc.TimerTail + 1) % tc.Capacity
+}
+func (tc *TimerCache) prev() {
+	tc.TimerTail = (tc.TimerTail - 1) % tc.Capacity
+}
+func (tc *TimerCache) Head() TimerCacheItem {
+	return tc.Values[tc.TimerHead]
+}
+
+func (c *Cache) createRenameEvent(oldNameEvent, newNameEvent Event, oldNameExist, newNameExist bool) {
+	if !newNameExist && oldNameExist { //ITEM REMOVED
+		oldNameEvent.Flags = ItemRemoved
```

Comment:
Do we care about overwriting any other flags on the event or is it safe to make this just a remove event?

Reply:
No the Remove operation is pretty the same for  all the file system types.